### DISCUSSION TITLE
API 5 JSON on Develop - album method

### DISCUSSION THREAD

**Discussion by Andrew-McGee** [ampache/ampache#2840] [General]
Hi folks,

I have my own frontend under development and I just switched to the latest develop branch using the new API. 
Some of the changes in returned data structures caught me out. Specifically the album method has added the 'album' element which is in the API 5 examples but it also seems to have slipped in another element [0] under that one before the usual data. 

So this is my old code that worked for me under 4:
 `$albm_results['artist']['id']`

This is what I read in the documentation examples:
`$albm_results['album']['artist']['id']`

But this is what works for me (after pulling some of my hair out):
`$albm_results['album'][0]['artist']['id']`

It looks like it is set in case album returns more than one result but I thought this call was only for a single album.

I also noticed the example for this call lists ['disk'] as an element but in my testing the returned element is ['diskcount'].

Happy to help update any docs as I work through these changes but just wanted to call out what I am finding.

*Note: It could also be me - nested multidimensional arrays really do my brain in sometimes! :)

**Comment by mitchray** [comment]
There has definitely been a conversation around this recently but I can't seem to find it... 

Side note, foam looks great!

**Reply by Andrew-McGee** [reply]
Thanks - it's getting there... slowly.

**Comment by Andrew-McGee** [comment]
Another one I will add here rather than start a new thread.

Playlists method returns in this format: 
`$results['playlist'][$cnt]['name']`

While the playlist (singular) method still uses:
`$results['name']`

Although the docs don't reflect this and it doesn't seem consistent without the [playlist] array returned.

**Comment by Andrew-McGee** [comment]
One more: The time element of the album method always seems to return zero. I checked the db and albums still definitely have a time field with non-zero value.

eg: Always returns 0:
`$results['album'][0]['time']`

**Comment by usox** [comment]
Well, tbh I'm not that happy with the state of the api, too. I spend several days creating unittests for the complete api (almost 100 methods) and therefor also tested every method manually. Although I'm quite happy about having all those tests now, it's wasn't fun to work with the api.

The problems from my pov:
- Some methods return way too much data (e.g. querying a list of albums should not contain songs, not even optional...)
- Non-deterministic result structure (sometimes with a key like `album`, sometimes not; lists, dicts, list of lists, ...)
- Optional arguments with non-deterministic defaults
- Requests/Results are hard to verify (json schema any one?)
- Results must fit into both json and xml structures
- Not versionable
- Results hard to cache in frontend applications/not cachable
- Errors are way too generic

**Reply by Andrew-McGee** [reply]
I agree with your points. The inconsistency is what I keep finding. I am happy to live with quirks as long as they are standardised! ;)

Also I would love to get sorted results returned (eg. reverse alphabetical order or something like that). Unless I juts haven't found how to do this yet.

**Comment by lachlan-00** [comment]
You are correct in that anything returning a single item (like album) should be only returning the data without the album element.

API5 isn't finished yet and i've made a few changes between draft 1 and now. I'll need to go through and update these docs and check the methods.

Please feel free to make an issue for these and put in any calls/returns you make and i'll get onto them tomorrow. I really want API5 to be a lot better than 4 by following the rules/standards we've set. (which is hard as they've never really existed before last year.)

**Comment by lachlan-00** [comment]
I see what that is, @usox the encode value needs to be false in the albums method. it's true by default so it's returning the album as an array
![image](https://user-images.githubusercontent.com/1305249/111923446-fdcd4280-8aea-11eb-9637-12227bcc5145.png)

**Reply by usox** [reply]
@lachlan-00 Yep, got it covered.

**Comment by usox** [comment]
@Andrew-McGee The result of the album method should now behave as expected. a00393fd4274d45e5d75ab6b09e6bb1f24873b33

**Reply by Andrew-McGee** [reply]
Thanks mate - appreciate it!
